core/domain/exp_services_test.py

```diff
@@ -881,13 +881,132 @@ def test_export_by_versions(self):
+        'cmd': exp_domain.CMD_EDIT_STATE_PROPERTY,
         'state_name': state_name,
         'property_name': property_name,
         'new_value': new_value
     }]
 
 
+class UpdateGadgetTests(ExplorationServicesUnitTests):
+    """Test updating a gadget."""
+
+    def setUp(self):
+        super(UpdateGadgetTests, self).setUp()
+        exploration = self.save_new_valid_exploration(
+            self.EXP_ID, self.owner_id)
+        # Default outcome specification for an interaction.
+        self.init_state_name = exploration.init_state_name
+
+        self.interaction_default_outcome = {
+            'dest': self.init_state_name,
+            'feedback': ['Incorrect', '<b>Wrong answer</b>'],
+            'param_changes': []
+        }
+
+    def test_add_gadget_cmd(self):
+        """test adding of a gadget"""
+        gadget_customization = {
+            'adviceObjects': {
+                'value': [{
+                    'adviceTitle': 'b',
+                    'adviceHtml': '<p>c</p>'
+                }]
+            }
+        }
+        exploration = exp_services.get_exploration_by_id(self.EXP_ID)
+
+        self.assertNotIn('NewGadget', exploration.get_all_gadget_names())
+        exp_services.update_exploration(self.owner_id, self.EXP_ID, [{
+            'cmd': exp_domain.CMD_ADD_GADGET,
+            'panel': 'bottom',
+            'gadget_dict' : {
+                'gadget_type': 'ScoreBar',
+                'gadget_name': 'NewGadget',
+                'customization_args': gadget_customization,
+                'visible_in_states':[feconf.DEFAULT_INIT_STATE_NAME]}
+            }], 'add a new gadget')
+
+        exploration = exp_services.get_exploration_by_id(self.EXP_ID)
+
+        self.assertIn('NewGadget', exploration.get_all_gadget_names())
+
+    def test_rename_gadget_cmd(self):
+        """test rename"""
+        gadget_customization = {
+            'adviceObjects': {
+                'value': [{
+                    'adviceTitle': 'b',
+                    'adviceHtml': '<p>c</p>'
+                }]
+            }
+        }
+
+        exploration = exp_services.get_exploration_by_id(self.EXP_ID)
+
+        exp_services.update_exploration(self.owner_id, self.EXP_ID, [{
+            'cmd': exp_domain.CMD_ADD_GADGET,
+            'panel': 'bottom',
+            'gadget_dict' : {
+                'gadget_type': 'ScoreBar',
+                'gadget_name': 'NewGadget',
+                'customization_args': gadget_customization,
+                'visible_in_states':[feconf.DEFAULT_INIT_STATE_NAME]}
+            }], 'add a new gadget')
+
+        exploration = exp_services.get_exploration_by_id(self.EXP_ID)
+
+        self.assertIn('NewGadget', exploration.get_all_gadget_names())
+        self.assertNotIn('gadget', exploration.get_all_gadget_names())
+
+        exp_services.update_exploration(self.owner_id, self.EXP_ID, [{
+            'cmd': exp_domain.CMD_RENAME_GADGET,
+            'new_gadget_name': 'gadget',
```

Comment:
Possibly a new name that's more consistent might be ``RenamedGadget``. This also helps to self-document your assertions. I also recommend you use the name ``Gadget`` instead of ``NewGadget`` in this test. ``NewGadget`` makes sense when testing whether a new gadget was added, but is a bit confusing when testing for renaming. It's quite clear to go from ``Gadget`` to ``RenamedGadget``, but ``NewGadget`` to anything else is a bit confusing.